### DISCUSSION TITLE
Photoprism is able to use TensorFlow on Raspberry PI 3/4

### DISCUSSION THREAD

**Discussion by l4nj1n9** [immich-app/immich#217] [General]
You wrote in your README.md:

> Raspberry Pi can be used but microservices container has to be comment out in docker-compose since TensorFlow has not been supported in Docker image on arm64v7 yet

I currently use Photoprism on my Raspberry Pi 3 and object / face recognition is working quite well. TensorFlow works. I just followed the instructions found [here](https://docs.photoprism.app/getting-started/raspberry-pi/#raspberry-pi-os). So I think you should inspire you from it to make Immich able to use TensorFlow on Raspberry Pi 3/4. What do you think ?

**Comment by alextran1502** [comment]
Thank you, I will take a look how they build their image for Arm64 and find a way to incorporate that into Immich 👍

**Comment by l4nj1n9** [comment]
@alextran1502 So, is Photoprism able to use TensorFlow on Raspberry Pi now ?

**Reply by alextran1502** [reply]
I don't know, you have to ask the Photoprism team

**Comment by l4nj1n9** [comment]
@alextran1502 Sorry, I'm quite tired those days. I meant : is Immish able to use TensorFlow on Raspberry Pi now ?

**Reply by alextran1502** [reply]
Haha no problem, we don't use TensorFlow anymore, we switch to Pytorch so you can now run it on the Pi. But it is better to offload the machine learning to another pc.

**Reply by l4nj1n9** [reply]
@alextran1502 Great, thank you !

**Reply by Eirikr70** [reply]
Hello, how would one offload the machine learning to another PC ? Should one install the dedicated container on the other PC and set up a network ?...

**Reply by alextran1502** [reply]
@Eirikr70 https://immich.app/docs/guides/machine-learning

**Reply by Eirikr70** [reply]
@alextran1502 Thanks. I hadn't been looking for it in that place.

**Reply by boboche** [reply]
with 1.79.1 I don't see any tagging on images, I see the exif, I see that the face detection works, but no tags or classification even if I change the model in the config.  Same issue with Rpi 4/4GB with SDimage or on a fresh ubuntu 22 VM.   Am I missing something or its broken?

**Reply by alextran1502** [reply]
@boboche We don't use image classification's results for anything actually, and we don't use traditional object detection methods. We use CLIP search in order to search for things and the information is encoded as an array of floating number.

**Reply by boboche** [reply]
Thanks for the reply Alex, I am trying to figure out the intended workflow and see how to fit my needs within what's available.  I see object tagging / image classification in this context: I have over a million images, I want to dump them on an offline solution that can help search/segmentation based off what is in the image or its context  (i.e. industrial, beach, buildings, sky, clouds, and what not).  I want to see the labels/tags or whatnot that are applied to the image in order to assess the tagging performance (i.e. calibrate it to 0.9, or 0.95/0.8 depending on  the signal to noise ratio) and finally, be able to see those tags will help me search relevant pictures and/or further filtering down to specifics.

That on top of all classification by date, ISO, lens and exif whatnots. 

The closest I've seen doing it properly is librephotos so far but immich is more mature and polished, My use case might be a niche scenario but aside from the very large quantity of images, I don't see how people wouldn't benefit from this if using a good AI model. 

Another approach would be to tag the image and embed this in the jpg info but I am dealing with raw images as well, and there are no out of the box solutions for such a workflow, made simple, that I am aware of.

Nice work for the overall project btw :)

**Reply by jrasm91** [reply]
The model we use now is pretty inaccurate. You bring up some good ideas related to search and filtering though. We're in the process of refactoring search, so hopefully it will improve in the future. We do have plans to enable user supplied tags in the future as well.
